src/org/opensolaris/opengrok/web/api/error/GenericExceptionMapper.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ */
+package org.opensolaris.opengrok.web.api.error;
+
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class GenericExceptionMapper implements ExceptionMapper<Exception> {
+
+    @Override
+    public Response toResponse(final Exception e) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+                .entity(e.getMessage())
+                .type(MediaType.TEXT_PLAIN)
```

Comment:
I'm more for json and something like `{status: 500, message: <message>}`

Reply:
Status is already in the response. Why duplicate?

Reply:
I saw it elsewhere however status can be ommited but still I vote for json with the message

Reply:
So just `{message: <message>}`?

I can turn the whole exception to json which would produce something like this:
```json
{"cause":null,"stackTrace":[{"methodName":"get","fileName":"ConfigurationController.java","lineNumber":49,"className":"org.opensolaris.opengrok.web.api.v1.controller.ConfigurationController","nativeMethod":false},{"methodName":"invoke0","fileName":"NativeMethodAccessorImpl.java","lineNumber":-2,"className":"sun.reflect.NativeMethodAccessorImpl","nativeMethod":true},{"methodName":"invoke","fileName":"NativeMethodAccessorImpl.java","lineNumber":62,"className":"sun.reflect.NativeMethodAccessorImpl","nativeMethod":false},{"methodName":"invoke","fileName":"DelegatingMethodAccessorImpl.java","lineNumber":43,"className":"sun.reflect.DelegatingMethodAccessorImpl","nativeMethod":false},{"methodName":"invoke","fileName":"Method.java","lineNumber":498,"className":"java.lang.reflect.Method","nativeMethod":false},{"methodName":"lambda$static$0","fileName":"ResourceMethodInvocationHandlerFactory.java","lineNumber":76,"className":"org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory","nativeMethod":false},{"methodName":"run","fileName":"AbstractJavaResourceMethodDispatcher.java","lineNumber":148,"className":"org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1","nativeMethod":false},{"methodName":"invoke","fileName":"AbstractJavaResourceMethodDispatcher.java","lineNumber":191,"className":"org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher","nativeMethod":false},{"methodName":"doDispatch","fileName":"JavaResourceMethodDispatcherProvider.java","lineNumber":243,"className":"org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker","nativeMethod":false},{"methodName":"dispatch","fileName":"AbstractJavaResourceMethodDispatcher.java","lineNumber":103,"className":"org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher","nativeMethod":false},{"methodName":"invoke","fileName":"ResourceMethodInvoker.java","lineNumber":493,"className":"org.glassfish.jersey.server.model.ResourceMethodInvoker","nativeMethod":false},{"methodName":"apply","fileName":"ResourceMethodInvoker.java","lineNumber":415,"className":"org.glassfish.jersey.server.model.ResourceMethodInvoker","nativeMethod":false},{"methodName":"apply","fileName":"ResourceMethodInvoker.java","lineNumber":104,"className":"org.glassfish.jersey.server.model.ResourceMethodInvoker","nativeMethod":false},{"methodName":"run","fileName":"ServerRuntime.java","lineNumber":277,"className":"org.glassfish.jersey.server.ServerRuntime$1","nativeMethod":false},{"methodName":"call","fileName":"Errors.java","lineNumber":272,"className":"org.glassfish.jersey.internal.Errors$1","nativeMethod":false},{"methodName":"call","fileName":"Errors.java","lineNumber":268,"className":"org.glassfish.jersey.internal.Errors$1","nativeMethod":false},{"methodName":"process","fileName":"Errors.java","lineNumber":316,"className":"org.glassfish.jersey.internal.Errors","nativeMethod":false},{"methodName":"process","fileName":"Errors.java","lineNumber":298,"className":"org.glassfish.jersey.internal.Errors","nativeMethod":false},{"methodName":"process","fileName":"Errors.java","lineNumber":268,"className":"org.glassfish.jersey.internal.Errors","nativeMethod":false},{"methodName":"runInScope","fileName":"RequestScope.java","lineNumber":289,"className":"org.glassfish.jersey.process.internal.RequestScope","nativeMethod":false},{"methodName":"process","fileName":"ServerRuntime.java","lineNumber":256,"className":"org.glassfish.jersey.server.ServerRuntime","nativeMethod":false},{"methodName":"handle","fileName":"ApplicationHandler.java","lineNumber":703,"className":"org.glassfish.jersey.server.ApplicationHandler","nativeMethod":false},{"methodName":"serviceImpl","fileName":"WebComponent.java","lineNumber":416,"className":"org.glassfish.jersey.servlet.WebComponent","nativeMethod":false},{"methodName":"service","fileName":"WebComponent.java","lineNumber":370,"className":"org.glassfish.jersey.servlet.WebComponent","nativeMethod":false},{"methodName":"service","fileName":"ServletContainer.java","lineNumber":389,"className":"org.glassfish.jersey.servlet.ServletContainer","nativeMethod":false},{"methodName":"service","fileName":"ServletContainer.java","lineNumber":342,"className":"org.glassfish.jersey.servlet.ServletContainer","nativeMethod":false},{"methodName":"service","fileName":"ServletContainer.java","lineNumber":229,"className":"org.glassfish.jersey.servlet.ServletContainer","nativeMethod":false},{"methodName":"internalDoFilter","fileName":"ApplicationFilterChain.java","lineNumber":231,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ApplicationFilterChain.java","lineNumber":166,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"WsFilter.java","lineNumber":53,"className":"org.apache.tomcat.websocket.server.WsFilter","nativeMethod":false},{"methodName":"internalDoFilter","fileName":"ApplicationFilterChain.java","lineNumber":193,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ApplicationFilterChain.java","lineNumber":166,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ResponseHeaderFilter.java","lineNumber":41,"className":"org.opensolaris.opengrok.web.ResponseHeaderFilter","nativeMethod":false},{"methodName":"internalDoFilter","fileName":"ApplicationFilterChain.java","lineNumber":193,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ApplicationFilterChain.java","lineNumber":166,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"StatisticsFilter.java","lineNumber":55,"className":"org.opensolaris.opengrok.web.StatisticsFilter","nativeMethod":false},{"methodName":"internalDoFilter","fileName":"ApplicationFilterChain.java","lineNumber":193,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ApplicationFilterChain.java","lineNumber":166,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"AuthorizationFilter.java","lineNumber":86,"className":"org.opensolaris.opengrok.web.AuthorizationFilter","nativeMethod":false},{"methodName":"internalDoFilter","fileName":"ApplicationFilterChain.java","lineNumber":193,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"doFilter","fileName":"ApplicationFilterChain.java","lineNumber":166,"className":"org.apache.catalina.core.ApplicationFilterChain","nativeMethod":false},{"methodName":"invoke","fileName":"StandardWrapperValve.java","lineNumber":199,"className":"org.apache.catalina.core.StandardWrapperValve","nativeMethod":false},{"methodName":"invoke","fileName":"StandardContextValve.java","lineNumber":96,"className":"org.apache.catalina.core.StandardContextValve","nativeMethod":false},{"methodName":"invoke","fileName":"AuthenticatorBase.java","lineNumber":494,"className":"org.apache.catalina.authenticator.AuthenticatorBase","nativeMethod":false},{"methodName":"invoke","fileName":"StandardHostValve.java","lineNumber":139,"className":"org.apache.catalina.core.StandardHostValve","nativeMethod":false},{"methodName":"invoke","fileName":"ErrorReportValve.java","lineNumber":92,"className":"org.apache.catalina.valves.ErrorReportValve","nativeMethod":false},{"methodName":"invoke","fileName":"AbstractAccessLogValve.java","lineNumber":651,"className":"org.apache.catalina.valves.AbstractAccessLogValve","nativeMethod":false},{"methodName":"invoke","fileName":"StandardEngineValve.java","lineNumber":87,"className":"org.apache.catalina.core.StandardEngineValve","nativeMethod":false},{"methodName":"service","fileName":"CoyoteAdapter.java","lineNumber":343,"className":"org.apache.catalina.connector.CoyoteAdapter","nativeMethod":false},{"methodName":"service","fileName":"Http11Processor.java","lineNumber":412,"className":"org.apache.coyote.http11.Http11Processor","nativeMethod":false},{"methodName":"process","fileName":"AbstractProcessorLight.java","lineNumber":66,"className":"org.apache.coyote.AbstractProcessorLight","nativeMethod":false},{"methodName":"process","fileName":"AbstractProtocol.java","lineNumber":754,"className":"org.apache.coyote.AbstractProtocol$ConnectionHandler","nativeMethod":false},{"methodName":"doRun","fileName":"NioEndpoint.java","lineNumber":1385,"className":"org.apache.tomcat.util.net.NioEndpoint$SocketProcessor","nativeMethod":false},{"methodName":"run","fileName":"SocketProcessorBase.java","lineNumber":49,"className":"org.apache.tomcat.util.net.SocketProcessorBase","nativeMethod":false},{"methodName":"runWorker","fileName":"ThreadPoolExecutor.java","lineNumber":1149,"className":"java.util.concurrent.ThreadPoolExecutor","nativeMethod":false},{"methodName":"run","fileName":"ThreadPoolExecutor.java","lineNumber":624,"className":"java.util.concurrent.ThreadPoolExecutor$Worker","nativeMethod":false},{"methodName":"run","fileName":"TaskThread.java","lineNumber":61,"className":"org.apache.tomcat.util.threads.TaskThread$WrappingRunnable","nativeMethod":false},{"methodName":"run","fileName":"Thread.java","lineNumber":748,"className":"java.lang.Thread","nativeMethod":false}],"message":"test","localizedMessage":"test","suppressed":[]}
```

Reply:
will the stacktrace be in the log? then it's not necessary to include it in the response - also this is not safe if we will expose some parts of the api for the public use in the future and forget about this.

Reply:
Exception is not logged. Maybe I can add `RequestEventListener` as specified here: <https://stackoverflow.com/questions/23048212/jersey-how-to-log-all-exceptions-but-still-invoke-exceptionmappers>

Reply:
That would be nice to have at least unexpected exceptions being logged.

Reply:
I would prefer logging only unexpected exceptions -> e.g. simple `logger` in `GenericExceptionMapper`. If some other exception mappers would like to log then they can do it by themselves. 

I don't like the idea of logging all exceptions (for example those because of bad requests (`ValidationException`, ...)).

What do you think?

Still, the question still remains what to return, just `{"message": <msg>}`?

Reply:
> What do you think?

Yes, just unexpected which would result to 500 anyway.

> Still, the question still remains what to return, just {"message": <msg>}?

Yes, I think so.

Reply:
By the way, with this code some 404 errors results in 500 or 404 with this exception mapper? I'm sort of confused with the SO link.

Reply:
In 500 :D because it throws `NotFoundException` which is caught by this handler... I miss Spring so much... everything is easier there :D

Reply:
Aha :D But that's not desired. Can you tweak it like in the first answer on SO? Or any other way of course.

```java
    private int getStatusCode(Throwable exception)
    {
        if (exception instanceof WebApplicationException)
        {
            return ((WebApplicationException)exception).getResponse().getStatus();
        }

        return Response.Status.INTERNAL_SERVER_ERROR.getStatusCode();
    }
```

Reply:
I will try to create `WebApplicationExceptionMapper`... hopefully it will work as it should.